staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go

```diff
@@ -0,0 +1,105 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package explain
+
+import (
+	"path/filepath"
+	"strings"
+	"testing"
+
+	sptest "k8s.io/apimachinery/pkg/util/strategicpatch/testing"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/discovery"
+	cmdtesting "k8s.io/kubectl/pkg/cmd/testing"
+	"k8s.io/kubectl/pkg/util/openapi"
+)
+
+var (
+	fakeSchema        = sptest.Fake{Path: filepath.Join("..", "..", "..", "testdata", "openapi", "swagger.json")}
+	FakeOpenAPISchema = testOpenAPISchema{
+		OpenAPISchemaFn: func() (openapi.Resources, error) {
+			s, err := fakeSchema.OpenAPISchema()
+			if err != nil {
+				return nil, err
+			}
+			return openapi.NewOpenAPIData(s)
+		},
+		OpenAPIGetter: &fakeSchema,
+	}
+)
+
+type testOpenAPISchema struct {
+	OpenAPISchemaFn func() (openapi.Resources, error)
+	OpenAPIGetter   discovery.OpenAPISchemaInterface
+}
+
+func TestExplainInvalidArgs(t *testing.T) {
+	tf := cmdtesting.NewTestFactory()
+	defer tf.Cleanup()
+
+	opts := NewExplainOptions("kubectl", genericclioptions.NewTestIOStreamsDiscard())
+	cmd := NewCmdExplain("kubectl", tf, genericclioptions.NewTestIOStreamsDiscard())
+	err := opts.Complete(tf, cmd, []string{})
+	if err != nil {
+		t.Fatalf("unexpected error %v", err)
+	}
+
+	err = opts.Validate()
+	if err.Error() != "You must specify the type of resource to explain. Use \"kubectl api-resources\" for a complete list of supported resources.\n" {
+		t.Error("unexpected non-error")
+	}
+
+	err = opts.Complete(tf, cmd, []string{"resource1", "resource2"})
+	if err != nil {
+		t.Fatalf("unexpected error %v", err)
+	}
+
+	err = opts.Validate()
+	if err.Error() != "We accept only this format: explain RESOURCE\n" {
+		t.Error("unexpected non-error")
+	}
+}
+
+func TestExplain(t *testing.T) {
+	tf := cmdtesting.NewTestFactory()
+	defer tf.Cleanup()
+
+	tf.OpenAPISchemaFunc = FakeOpenAPISchema.OpenAPISchemaFn
+	tf.FakeOpenAPIGetter = FakeOpenAPISchema.OpenAPIGetter
+	tf.ClientConfigVal = cmdtesting.DefaultClientConfig()
+
+	ioStreams, _, buf, _ := genericclioptions.NewTestIOStreams()
+	cmd := NewCmdExplain("kubectl", tf, ioStreams)
+	cmd.Run(cmd, []string{"pods"})
+	if !strings.Contains(buf.String(), "KIND:     Pod") {
+		t.Fatalf("expected output should include pod kind")
+	}
+
+	cmd.Flags().Set("recursive", "true")
+	cmd.Run(cmd, []string{"pods"})
+	if !strings.Contains(buf.String(), "KIND:     Pod") ||
+		!strings.Contains(buf.String(), "apiVersion\t<string>") {
+		t.Fatalf("expected output should include pod kind")
+	}
+
+	cmd.Flags().Set("api-version", "v1")
+	cmd.Run(cmd, []string{"pods"})
```

Comment:
Can you use something that has multiple version defined in the Open API spec and make sure that it shows the correct version?

So for example, I see in `testdata/openapi/swagger.json` that `cronjobs` has both `batch/v1beta1` and `batch/v1` versions defined.

So maybe test it both ways one with `--api-versions=batch/v1beta1` and then with `--api-versions=batch/v1` and look for the second line where is shows like:
```
VERSION:  batch/v1
```